javascript/ql/src/Security/CWE-020/MissingOriginCheck.ql

```diff
@@ -0,0 +1,73 @@
+/**
+ * @name Missing origin verification in `postMessage` handler
+ * @description Missing origin verification in a `postMessage` handler allows any windows to send arbitrary data to the handler.
+ * @kind problem
+ * @problem.severity warning
+ * @security-severity 5
+ * @precision medium
+ * @id js/missing-origin-check
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-020
+ */
+
+import javascript
+
+/** A function that handles "message" events. */
+class PostMessageHandler extends DataFlow::FunctionNode {
+  override PostMessageEventHandler astNode;
+
+  /** Gets the parameter that contains the event. */
+  DataFlow::ParameterNode getEventParameter() {
+    result = DataFlow::parameterNode(astNode.getEventParameter())
+  }
+}
+
+/** Gets a reference to the event from a postmessage `handler` */
+DataFlow::SourceNode event(DataFlow::TypeTracker t, PostMessageHandler handler) {
+  t.start() and
+  result = handler.getEventParameter()
+  or
+  exists(DataFlow::TypeTracker t2 | result = event(t2, handler).track(t2, t))
+}
+
+/** Gets a reference to the .origin from a postmessage event. */
+DataFlow::SourceNode origin(DataFlow::TypeTracker t, PostMessageHandler handler) {
+  t.start() and
+  result = event(DataFlow::TypeTracker::end(), handler).getAPropertyRead("origin")
+  or
+  result =
+    origin(t.continue(), handler)
+        .getAMethodCall([
+            "toString", "toLowerCase", "toUpperCase", "toLocaleLowerCase", "toLocaleUpperCase"
```

Comment:
Not relevant for this PR:

> "toLocaleLowerCase", "toLocaleUpperCase"

Hmm. I would consider it a bug to use those instead of the canonical `to(Lower|Upper)Case` calls, but perhaps the programmers are saved by the fact that all browser origins normalizes the same way regardless of locales.

Reply:
> Hmm. I would consider it a bug to use those instead of the canonical to(Lower|Upper)Case calls 

It would definitely be weird to use the `Locale` variants of those methods.  
I mainly included them because Copilot suggested them.